addons/binding/org.openhab.binding.mcp23017/src/main/java/org/openhab/binding/mcp23017/handler/AbstractMcp23017Handler.java

```diff
@@ -0,0 +1,105 @@
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.mcp23017.handler;
+
+import static org.openhab.binding.mcp23017.Consts.*;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.eclipse.smarthome.config.core.Configuration;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.pi4j.gpio.extension.mcp.MCP23017GpioProvider;
+import com.pi4j.gpio.extension.mcp.MCP23017Pin;
+import com.pi4j.io.gpio.Pin;
+import com.pi4j.io.i2c.I2CBus;
+import com.pi4j.io.i2c.I2CFactory.UnsupportedBusNumberException;
+
+/**
+ * The {@link AbstractMcp23017Handler} is base class for MCP23017 chip support
+ *
+ * @author Anatol Ogorek
+ */
+public abstract class AbstractMcp23017Handler extends BaseThingHandler {
+
+    protected Logger logger = LoggerFactory.getLogger(getClass());
+    protected static final Map<Integer, MCP23017GpioProvider> mcpProviders = new HashMap<>();
+
+    protected Integer address = null;
+    protected Integer busNumber = null;
+    protected Pin pin = null;
+
+    /**
+     * the polling interval mcp23071 check interrupt register (optional, defaults to 50ms)
+     */
+    protected int pollingInterval = 50;
+
+    public AbstractMcp23017Handler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void initialize() {
+        checkConfiguration();
+        bindGpioPin(getThing().getUID().getAsString());
+
+    }
+
+    protected void checkConfiguration() {
+        Configuration configuration = getConfig();
+        address = Integer.parseInt((configuration.get(ADDRESS)).toString(), 16);
+        busNumber = Integer.parseInt((configuration.get(BUS_NUMBER)).toString());
+        try {
+            String pinCode = (String) configuration.get(PIN);
+            pin = (Pin) MCP23017Pin.class.getField("GPIO_" + pinCode).get(null);
+            updateStatus(ThingStatus.ONLINE);
+        } catch (IllegalArgumentException | IllegalAccessException | NoSuchFieldException | SecurityException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "An exception occurred while adding PIN. Check PIN configuration.");
+        }
+
+    }
+
+    @Override
+    public void handleRemoval() {
+        super.handleRemoval();
+    }
+
+    public MCP23017GpioProvider getProvider() {
+        MCP23017GpioProvider mcp = mcpProviders.get(address);
+        if (mcp == null) {
+            try {
+                mcp = new MCP23017GpioProvider(busNumber, address);
+                mcp.setPollingTime(pollingInterval);
+            } catch (UnsupportedBusNumberException | IOException ex) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
```

Comment:
Would the exception message contain useful information if so, please pass it to the `updateStatus`